pkg/cli/admin/upgrade/recommend/recommend.go

```diff
@@ -145,82 +138,138 @@ func (o *options) Run(ctx context.Context) error {
+			if conditionalUpdate.Release.Image == update.Image {
+				found = true
 				break
 			}
 		}
-		fmt.Fprintf(o.Out, "\nAdditional updates which are not recommended%s for your cluster configuration are available, to view those re-run the command with --include-not-recommended.\n", qualifier)
-	}
+		if found {
+			continue
+		}
 
-	// TODO: print previous versions
+		version, err := semver.Parse(update.Version)
+		if err != nil {
+			fmt.Fprintf(o.ErrOut, "warning: Cannot parse SemVer available update %q: %v", update.Version, err)
+		}
 
-	return nil
-}
+		if minorBuckets := majorMinorBuckets[version.Major]; minorBuckets == nil {
+			majorMinorBuckets[version.Major] = make(map[uint64][]configv1.ConditionalUpdate, 0)
+		}
 
-func containsNotRecommendedUpdate(updates []configv1.ConditionalUpdate) bool {
-	for _, update := range updates {
-		if c := findCondition(update.Conditions, "Recommended"); c != nil && c.Status != metav1.ConditionTrue {
-			return true
+		if minorBucket := majorMinorBuckets[version.Major][version.Minor]; minorBucket == nil {
+			majorMinorBuckets[version.Major][version.Minor] = []configv1.ConditionalUpdate{}
 		}
+
+		majorMinorBuckets[version.Major][version.Minor] = append(majorMinorBuckets[version.Major][version.Minor], configv1.ConditionalUpdate{
+			Release: cv.Status.AvailableUpdates[i],
+		})
 	}
-	return false
-}
 
-// sortReleasesBySemanticVersions sorts the input slice in decreasing order.
-func sortReleasesBySemanticVersions(versions []configv1.Release) {
-	sort.Slice(versions, func(i, j int) bool {
-		a, errA := semver.Parse(versions[i].Version)
-		b, errB := semver.Parse(versions[j].Version)
-		if errA == nil && errB != nil {
-			return true
+	if len(majorMinorBuckets) == 0 {
+		fmt.Fprintf(o.Out, "No updates available. You may still upgrade to a specific release image with --to-image or wait for new updates to be available.\n")
+		return nil
+	}
+
+	fmt.Fprintf(o.Out, "Updates:\n")
+	// set the minimal cell width to 14 to have a larger space between the columns for shorter versions
+	w := tabwriter.NewWriter(o.Out, 14, 2, 1, ' ', 0)
+	headerQueued := true
+	fmt.Fprintf(w, "  VERSION\tIMAGE\n")
+	// TODO: add metadata about version
+
+	majors := make([]uint64, 0, len(majorMinorBuckets))
+	for major := range majorMinorBuckets {
+		majors = append(majors, major)
+	}
+	sort.Slice(majors, func(i, j int) bool {
+		return majors[i] > majors[j] // sort descending, major updates bring lots of features (enough to justify breaking backwards compatibility)
+	})
+	lastWasLong := false
+	for _, major := range majors {
+		minors := make([]uint64, 0, len(majorMinorBuckets[major]))
+		for minor := range majorMinorBuckets[major] {
+			minors = append(minors, minor)
 		}
-		if errB == nil && errA != nil {
-			return false
+		sort.Slice(minors, func(i, j int) bool {
+			return minors[i] > minors[j] // sort descending, minor updates bring both feature and bugfixes
+		})
```

Comment:
```
minors := slices.SortedFunc(maps.Keys(majorMinorBuckets[major]), descSortCmp)
```

Reply:
I've pushed 46d8a1c096dc -> 07d7b03745e5b3 with the `SortedFunc` thing.

Reply:
ah, maybe we can't use Go 1.23 yet][1]:

```
# github.com/openshift/oc/pkg/cli/admin/upgrade/recommend
pkg/cli/admin/upgrade/recommend/recommend.go:200:20: undefined: slices.SortedFunc
pkg/cli/admin/upgrade/recommend/recommend.go:200:36: undefined: maps.Keys
```

I'll go back to what I had before...

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_oc/1889/pull-ci-openshift-oc-master-images/1842241799914000384#1:build-log.txt%3A69

Reply:
Rolled back to the old style in 07d7b037 -> d8cac253.